### DISCUSSION TITLE
Vertical Link objects unexpectedly protrude.

### DISCUSSION THREAD

**Discussion by jayfella** [dyn4j/dyn4j#302] [Q&A] [answered]
I have made some small modifications to the LinkTerrain sample (below) that adds vertical links to simulate a join (e.g. chunks). When the simulation is run you can see that it "trips over" the vertical links, even though the top point is in exactly the same location as the horizontal links. This collision is clearly represented with a listener with the collision normal as (0,1).

If you toggle `useEpsilon` - which reduces the height of the vertical links we can see there's a margin of error that can be calculated.

I would expect, though, that since the vertical links are not protruding that they would not interact with the moving body since the length of penetration should be zero. Is there any way we can avoid this?

```java
package org.dyn4j.samples;

import java.util.List;

import org.dyn4j.collision.narrowphase.Sat;
import org.dyn4j.geometry.Geometry;
import org.dyn4j.geometry.Link;
import org.dyn4j.geometry.MassType;
import org.dyn4j.geometry.Vector2;
import org.dyn4j.samples.framework.Camera;
import org.dyn4j.samples.framework.SimulationBody;
import org.dyn4j.samples.framework.SimulationFrame;

/**
 * A simple scene of a terrain made using the {@link Link}s to avoid
 * the internal-edge collision problem.
 *
 * @author William Bittle
 * @version 5.0.0
 * @since 3.2.2
 */
public class TerrainSticking extends SimulationFrame {
    /**
     * The serial version id
     */
    private static final long serialVersionUID = -3675099977835892473L;

    /**
     * Default constructor for the window
     */
    public TerrainSticking() {
        super("Terrain Sticking");

        this.pause();
    }

    /**
     * Creates game objects and adds them to the world.
     */
    protected void initializeWorld() {
        this.world.setNarrowphaseDetector(new Sat());

        // the terrain
        List<Link> links = Geometry.createLinks(
                new Vector2[]{
                        new Vector2(6.0, -0.5),
                        new Vector2(5.5, -0.5),
                        new Vector2(5.0, -0.5),
                        new Vector2(4.5, -0.5),
                        new Vector2(4.0, -0.5),
                        new Vector2(3.5, -0.5),
                        new Vector2(3.0, -0.5),
                        new Vector2(2.5, -0.5),
                        new Vector2(2.0, -0.5),
                        new Vector2(1.5, -0.5),
                        new Vector2(1.0, -0.5),
                        new Vector2(0.5, -0.5),
                        new Vector2(0.0, -0.5),

                        new Vector2(-0.5, -0.5),
                        new Vector2(-1.0, -0.5),
                        new Vector2(-1.5, -0.5),
                        new Vector2(-2.0, -0.5),
                        new Vector2(-2.5, -0.5),
                        new Vector2(-3.0, -0.5),
                        new Vector2(-3.5, -0.5),
                        new Vector2(-4.0, -0.5),
                        new Vector2(-4.5, -0.5),
                        new Vector2(-5.0, -0.5),
                        new Vector2(-5.5, -0.5),
                        new Vector2(-6.0, -0.5),
                }, false);


        for (int i = 0; i < links.size(); i++) {
            Link horizontal = links.get(i);
            horizontal.setUserData("HORIZONTAL");

            SimulationBody floor = new SimulationBody();

            floor.addFixture(horizontal);

            if (i % 4 == 0) {
                double epsilon = 0.03;
                boolean useEpsilon = false;

                Link vertical = new Link(horizontal.getVertices()[0].copy().subtract(0, useEpsilon ? epsilon : 0),
                         horizontal.getVertices()[0].copy().subtract(0, 1));

                horizontal.setUserData("VERTICAL");

                floor.addFixture(vertical);
            }

            floor.setMass(MassType.INFINITE);

            this.world.addBody(floor);
        }

        // the body
        SimulationBody slider = new SimulationBody();
        slider.addFixture(Geometry.createSquare(0.25));
        slider.setMass(MassType.NORMAL);
        slider.setLinearVelocity(16.6, 0);
        slider.translate(-5.5, -0.2);
        this.world.addBody(slider);
    }

    /* (non-Javadoc)
     * @see com.jayfella.physics.dyn4j.samples.framework.SimulationFrame#initializeCamera(com.jayfella.physics.dyn4j.samples.framework.Camera)
     */
    @Override
    protected void initializeCamera(Camera camera) {
        super.initializeCamera(camera);
        camera.scale = 64.0;
    }

    /**
     * Entry point for the example application.
     *
     * @param args command line arguments
     */
    public static void main(String[] args) {
        TerrainSticking simulation = new TerrainSticking();
        simulation.run();
    }
}
```

**Comment by wnbittle** [comment, marked as answer]
I'm not at a place to test this, but I think the issue is with how you are constructing the vertical links.

A `Link` by nature doesn't act any different compared to a `Segment` unless the you also supply adjacency information.  The adjacency information is used to determine what the corrected normal should be, so when it doesn't have that, it basically works like a `Segment`.  The `Link` class has a four argument constructor that can provide the adjacency information.

That said, and the part that will be tricky, is when you use multiple sets of `Link`s.  There's no normal correction between them either for the same reason - with no adjacency information, the correction can't be determined.

Maybe with this information the behavior will be clearer.  If you could describe the desired effect or simulation, I or others might be able to provide another solution or a way to adjust your current solution to do what you are looking for.

**Comment by jayfella** [comment]
The information about adjacent links was the information I needed, and thank you for the explanation. My scenario is creating shapes from a grid whose values can be 1 (solid) or 0 (air). For example:

![image](https://github.com/user-attachments/assets/cb818672-d3c9-4033-a3ed-00e8bb9acfc3)

I ended up solving the problem by using the links to draw around the perimeter of the shape. Now that they are "joined" with the adjacency information it works a lot more in-line with my expectations.

Appreciate the help @wnbittle - and thanks for all of your hard work on the library itself.

**Comment by wnbittle** [comment]
Glad to hear it and thank you for posting more detail on the problem and solution so others can benefit!
